compiler/rustc_privacy/src/lib.rs

```diff
@@ -413,50 +429,65 @@ struct EmbargoVisitor<'tcx> {
     ///     n::p::f()
     /// }
     macro_reachable: FxHashSet<(LocalDefId, LocalDefId)>,
-    /// Previous visibility level; `None` means unreachable.
-    prev_level: Option<Level>,
     /// Has something changed in the level map?
     changed: bool,
 }
 
 struct ReachEverythingInTheInterfaceVisitor<'a, 'tcx> {
-    level: Option<Level>,
+    effective_vis: Option<EffectiveVisibility>,
     item_def_id: LocalDefId,
     ev: &'a mut EmbargoVisitor<'tcx>,
 }
 
 impl<'tcx> EmbargoVisitor<'tcx> {
-    fn get(&self, def_id: LocalDefId) -> Option<Level> {
-        self.effective_visibilities.public_at_level(def_id)
+    fn get(&self, def_id: LocalDefId) -> Option<EffectiveVisibility> {
+        self.effective_visibilities.effective_vis(def_id).copied()
     }
 
     /// Updates node level and returns the updated level.
-    fn update(&mut self, def_id: LocalDefId, level: Option<Level>) -> Option<Level> {
-        let old_level = self.get(def_id);
-        // Visibility levels can only grow.
-        if level > old_level {
-            self.effective_visibilities.set_public_at_level(
-                def_id,
-                || ty::Visibility::Restricted(self.tcx.parent_module_from_def_id(def_id)),
-                level.unwrap(),
-            );
-            self.changed = true;
-            level
-        } else {
-            old_level
+    fn update(
+        &mut self,
+        def_id: LocalDefId,
+        item_effective_vis: Option<EffectiveVisibility>,
+        level: Level,
+    ) {
+        let nominal_vis = self.tcx.local_visibility(def_id);
+        self.calculate(def_id, item_effective_vis, Some(nominal_vis), level);
+    }
+
+    fn calculate(
+        &mut self,
+        def_id: LocalDefId,
+        item_effective_vis: Option<EffectiveVisibility>,
+        nominal_vis: Option<ty::Visibility>,
+        level: Level,
+    ) {
+        let private_vis = ty::Visibility::Restricted(self.tcx.parent_module_from_def_id(def_id));
+
+        if Some(private_vis) != nominal_vis {
```

Comment:
It may be cheaper to not do this optimization, but to evaluate `private_vis` lazily instead (at least it was like this in `rustc_resolve`).
But it needs measuring.

Reply:
I'll check this myself later.

Reply:
This suggestion didn't help, https://github.com/rust-lang/rust/pull/110907#issuecomment-1534013288 is slightly slower than https://perf.rust-lang.org/compare.html?start=82cd953c7c43e64dae6f705ce2f07b291f0e22e3&end=869bc6f4741af66a11a05a0ce17235fd73041021&stat=instructions:u.